libtenzir/builtins/contexts/geoip.cpp

```diff
@@ -0,0 +1,485 @@
+      }
+      return caf::make_error(ec::invalid_argument,
+                             fmt::format("invalid argument `{}`", key));
+    }
+    if (not field_name) {
+      return caf::make_error(ec::invalid_argument, "missing argument `field`");
+    }
+    auto field_builder = series_builder{};
+    auto column_offset
+      = caf::get<record_type>(slice.schema()).resolve_key(*field_name);
+    if (not column_offset) {
+      for (auto i = size_t{0}; i < slice.rows(); ++i) {
+        field_builder.null();
+      }
+      return field_builder.finish();
+    }
+    auto [slice_type, slice_array] = column_offset->get(resolved_slice);
+    if (slice_type != type{ip_type{}} and slice_type != type{string_type{}}) {
+      // No ip type = no enrichment.
+      field_builder.null();
+      return field_builder.finish();
+    }
+    for (const auto& value : values(slice_type, *slice_array)) {
+      auto address_info_error = 0;
+      auto ip_string = fmt::to_string(value);
+      if (slice_type == type{string_type{}}) {
+        // Unquote IP strings.
+        ip_string.erase(0, 1);
+        ip_string.erase(ip_string.size() - 1);
+      }
+      auto result = MMDB_lookup_string(&mmdb, ip_string.data(),
+                                       &address_info_error, &status);
+      if (address_info_error != MMDB_SUCCESS) {
+        return caf::make_error(
+          ec::lookup_error,
+          fmt::format("error looking up IP address '{}' in "
+                      "GeoIP database: {}",
+                      ip_string, gai_strerror(address_info_error)));
+      }
+      if (status != MMDB_SUCCESS) {
+        return caf::make_error(
+          ec::lookup_error, fmt::format("error looking up IP address '{}' in "
+                                        "GeoIP database: {}",
+                                        ip_string, MMDB_strerror(status)));
+      }
+      if (result.found_entry) {
+        status = MMDB_get_entry_data_list(&result.entry, &entry_data_list);
+        if (status != MMDB_SUCCESS) {
+          return caf::make_error(
+            ec::lookup_error, fmt::format("error looking up IP address '{}' in "
+                                          "GeoIP database: {}",
+                                          ip_string, MMDB_strerror(status)));
+        }
+        auto* entry_data_list_it = entry_data_list;
+        auto output = record{};
+        entry_data_list_it
+          = entry_data_list_to_record(entry_data_list_it, &status, output);
+        if (status != MMDB_SUCCESS) {
+          return caf::make_error(
+            ec::lookup_error, fmt::format("error looking up IP address '{}' in "
+                                          "GeoIP database: {}",
+                                          ip_string, MMDB_strerror(status)));
+        }
+        auto r = field_builder.record();
+        r.field("key", value);
+        r.field("context", output);
+        r.field("timestamp", std::chrono::system_clock::now());
```

Comment:
Are these three relevant here, or should it just be `output`?

Reply:
1. `key` is the IP that matches against the specified field in `enrich`, I think it should stay there. Imagine an enriched event that contains multiple IP address values. There is no guarantee that `context` contains the IP address. How would you, by only looking at the enrichment without the `key` value, deduce again which IP value the enrichment was matched against?
Also, am not 100% confident we could just insert something like an `"address": value` data field into the `output` record without possibly overwriting anything.
The name `key` to me doesn't make sense though, I changed it to `address` in https://github.com/tenzir/tenzir/pull/3731/commits/bbeaf8fd7b7a06a2795ea3f921a7a317b4b99150.
2. `context` is obvious.
3. I think we specified that all hits should have a timestamp. Was that only for the `lookup-table`?